examples/common/convert_c4.py

```diff
@@ -4,105 +4,359 @@
+
+
+class ConcatC4(ShardedC4):
 
-    return args.parse_args()
+    def __init__(self, sep_text: str, *args, **kwargs):
+        self.sep_text = sep_text
+        super().__init__(*args, **kwargs)
 
+    def generate_samples(self) -> Iterable[Dict[str, bytes]]:
+        """Generator over each dataset sample.
 
-def build_hf_c4_dataset(split: str) -> IterableDataset:
+        Args:
+            samples (IterableDataset): An iterable dataset that is multi-worker compatible
+
+        Yields:
+            Sample dicts.
+        """
+
+        n_samples = 0
+        buffer = {}
+        while True:
+            for batch in self.loader:
+                for sample in batch['text']:
+                    buffer['text'] = buffer.get('text', '') + self.sep_text + sample
+                    while len(buffer['text']) >= self.max_length:
+                        concat_sample = {}
+                        concat_sample['text'] = buffer['text'][:self.
+                                                                max_length]
+                        buffer['text'] = buffer['text'][self.max_length:]
+
+                        yield {
+                            'text': concat_sample['text'].encode('utf-8')
+                        }
+                        n_samples += 1
+                        if n_samples == self.expected_num_samples:
+                            return
+
+
+class TokenizedC4(ShardedC4):
+    """A tokenized, concatenated, iterable dataset.
+
+    To use data created by this class and written to MDS format:
+
+    ```python
+        import torch
+        from streaming.base import StreamingDataset
+        from transformers import AutoTokenizer
+
+        tokenizer = AutoTokenizer.from_pretrained('your/tokenizer')
+        ds = StreamingDataset(local='mds-data-folder', split='val')
+
+        # note, you need to copy the numpy array because the original is non-writeable
+        # and torch does not support non-writeable tensors, so you get a scary warning and
+        # if you do try to write to the tensor you get undefined behavior
+        tokens = torch.from_numpy(np.frombuffer(ds[0]['tokens'], dtype=np.int64).copy())
+        print(tokenizer.decode(tokens))
+    ```
+    """
+
+    def __init__(self, tokenizer: PreTrainedTokenizerBase, *args, **kwargs):
+        self.tokenizer = tokenizer
+        super().__init__(*args, **kwargs)
+
+    def generate_samples(self) -> Iterable[Dict[str, bytes]]:
+        """Generator over each dataset sample.
+
+        Args:
+            samples (IterableDataset): An iterable dataset that is multi-worker compatible
+
+        Yields:
+            Sample dicts.
+        """
+
+        n_samples = 0
+        buffer = {}
+        while True:
+            for batch in self.loader:
+                encoded = self.tokenizer(batch['text'], truncation=False, padding=False)
+                for iids in encoded['input_ids']:
+                    buffer['tokens'] = buffer.get('tokens', []) + iids
```

Comment:
should there be a sep token here too?

Reply:
Reasonable question. Unless you have strong feelings otherwise, I think not.

We _do_ end up with sequences that start without a BOS token... but those are the sequences that start not at the beginning of a true sequence.